2019-12-naming-part1/naming.md

```diff
@@ -0,0 +1,231 @@
+
+You might determine that the cost of implementing the new requirement is too high in the present and poses too much risk in the near future. You'd prefer to prevent the change altogether. When you push back on the product team, explaining that it's too complex and expensive to make now, expect looks of befuddlement in response. Those looks are justified. A simple request should be simple to implement. The more you can harmonize the seemingly simple and the actually simple, the more you engender organizational trust in software engineering.
+
+This might seem like hyperbole. How can a single poorly named function make your programming life unpleasant and engender organizational mistrust in software engineering? A single poorly named function probably won't. But, the *habit* of naming functions after their current implementation will infect your codebase, making all aspects of development slower, more expensive, and less fun.
+
+Function names should tell the right story. Good names are resilient &mdash; they stand the test of time through changes in implementation. Naming is difficult, but names matter. It's incumbent upon you to develop this core programming skill. So, how should you name functions? 
+
+`trim_whitespace` is clearly named after its current implementation. The requirement to reorder the name to `last, first` order has forced you into a decision: to rename or not to rename. But, that decision could have been avoided altogether, by wresting your mind from granular implementation details to think about names more categorically. Your function name should move one rung higher.
+
+On the Abstraction Ladder, `trim_whitespace` occupies a lower rung. It's literal and concrete. In order to move the function name one rung higher, you must consider the function from the perspective of its callers. What service is the function providing?
+
+One way to derive names is to consider other things in the same category as the current implementation and then to name that category. What other things might you do to a name to present it? A simple matrix helps:
+
+| Person name                  | ???               |
+|------------------------------|-------------------|
+| Remove extraneous whitespace | `trim_whitespace` |
+| Present last name first      | `last_first`      |
+| Capitalize words             | `capitalize`      |
+| Prepend salutation           | `add_salutation`  |
+
+The verb `present` is a viable candidate for the missing column heading in the matrix above. But the name of the module is `PersonPresenter`. Using the word `present` would read like an echo chamber: `PersonPresenter.present_name`.
```

Comment:
Remember that the name of the module is `PersonPresenter` so naming the function `present` would read like an echo chamber: `PersonPresenter.present_name`.

Reply:
Resolved in 345a1339f8f2b4e506e3897474916427967dd070